src/ptz-device.cpp

```diff
@@ -351,6 +351,14 @@ obs_data_array_t *ptz_devices_get_config()
 	return ptzDeviceList.getConfigs();
 }
 
+const char *ptz_device_get_name(uint32_t device_id)
+{
+	PTZDevice *ptz = ptzDeviceList.getDevice(device_id);
+	if (ptz)
+		return QT_TO_UTF8(ptz->objectName());
+	return NULL;
+}
+
```

Comment:
There is a scoping problem with this implementation. QT_TO_UTF8() creates a transitory `const char *` string on the stack, but that object is destroyed on leaving the function. So the pointer returned no longer points to a valid object when the caller tries to use it. /Sometimes/ it will look like everything works if nothing has yet overwritten that part of the stack. On MacOS (Arm) it isn't working.

The safe thing to do is keep the `const char *` reference contained entirely within the function using it. It is okay to pass it into other functions, but not to return it to the caller. To get it working I reworked the code to look up the obs_source_t in the same helper. Here are the changes I made against your patch to get it working:

```
diff --git a/src/ptz-action-source.c b/src/ptz-action-source.c
index 0ff562d..8b1a672 100644
--- a/src/ptz-action-source.c
+++ b/src/ptz-action-source.c
@@ -104,20 +104,15 @@ static bool is_ptz_in_preview(struct ptz_action_source_data *context)

 static bool is_ptz_device_id_active_in_program(uint32_t device_id)
 {
-       const char *ptzname = ptz_device_get_name(device_id);
-       // find source with same name of the ptz
-       obs_source_t *cam_source = obs_get_source_by_name(ptzname);
-       if (cam_source) {
-               obs_source_t *program = obs_frontend_get_current_scene();
-               bool ptz_in_use =
-                       ptz_scene_is_source_active(program, cam_source);
-               //blog(LOG_INFO, "PREVIEW ACTION: is ptz %s (%d) in program? %s",
-               //     ptzname, device_id, ptz_in_use ? "true" : "false");
-               obs_source_release(program);
-               obs_source_release(cam_source);
-               return ptz_in_use;
-       }
-       return false;
+       obs_source_t *cam_source = ptz_device_get_source(device_id);
+       if (!cam_source)
+               return false;
+
+       obs_source_t *program = obs_frontend_get_current_scene();
+       bool ptz_in_use = ptz_scene_is_source_active(program, cam_source);
+       obs_source_release(program);
+       obs_source_release(cam_source);
+       return ptz_in_use;
 }

 static void ptz_action_source_fe_callback(enum obs_frontend_event event,
diff --git a/src/ptz-device.cpp b/src/ptz-device.cpp
index 116441c..4f5f11f 100644
--- a/src/ptz-device.cpp
+++ b/src/ptz-device.cpp
@@ -351,12 +351,12 @@ obs_data_array_t *ptz_devices_get_config()
        return ptzDeviceList.getConfigs();
 }

-const char *ptz_device_get_name(uint32_t device_id)
+obs_source_t *ptz_device_get_source(uint32_t device_id)
 {
        PTZDevice *ptz = ptzDeviceList.getDevice(device_id);
-       if (ptz)
-               return QT_TO_UTF8(ptz->objectName());
-       return NULL;
+       if (!ptz)
+               return NULL;
+       return obs_get_source_by_name(QT_TO_UTF8(ptz->objectName()));
 }

 void ptz_devices_set_config(obs_data_array_t *devices)
diff --git a/src/ptz.h b/src/ptz.h
index 2737bfa..c20cec7 100644
--- a/src/ptz.h
+++ b/src/ptz.h
@@ -27,7 +27,7 @@ extern void ptz_load_controls(void);
 extern void ptz_load_settings(void);

 extern obs_data_array_t *ptz_devices_get_config(void);
-extern const char *ptz_device_get_name(uint32_t device_id);
+extern obs_source_t *ptz_device_get_source(uint32_t device_id);
 extern void ptz_devices_set_config(obs_data_array_t *devices);

 extern bool ptz_scene_is_source_active(obs_source_t *scene,
```

Reply:
Great! I couldn't find anything about QT_TO_UTF8 on google, I was expecting only some kind of type conversion because of the upper-case name...